File: gastric_gland_analyser/_population_factory.py

from .population import Population
from .cell import Cell

class _PopulationFactory:
    def __init__(self, timepoint:float, node_locations, boundary_nodes, location_indices):
        self.node_locations = node_locations
        self.boundary_nodes = boundary_nodes
        self.location_indices = location_indices
        self.timepoint = timepoint

        self.n_nodes = len(node_locations)
        assert self.n_nodes == len(self.boundary_nodes)
        self.n_real_cells = len(self.location_indices)

        self.cell_ages = None
        self.cell_ancestors = None
        self.cell_areas = None
        self.cell_types = None
        self.cell_proliferative_types = None


    def create_population(self):
        population = Population()
        population.cells = [Cell() for _ in range(self.n_real_cells)]

        for i, location_index in enumerate(self.location_indices):
            population.cells[i].location = self.node_locations[location_index]
            population.cells[i].location_index = location_index
            assert self.boundary_nodes[location_index] == 0

        if self.cell_ages is not None:
            for i, age in enumerate(self.cell_ages):
                population.cells[i].age = age
        
        if self.cell_ancestors is not None:
            for i, ancestor in enumerate(self.cell_ancestors):
                population.cells[i].ancestor = ancestor
        
        if self.cell_areas is not None:
            for i, area in enumerate(self.cell_areas):
                population.cells[i].area = area

        if self.cell_types is not None:
            for i, cell_type in enumerate(self.cell_types):
                population.cells[i].cell_type = cell_type
        
        if self.cell_proliferative_types is not None:
            for i, proliferative_cell_type in enumerate(self.cell_proliferative_types):
                population.cells[i].proliferative_cell_type = proliferative_cell_type
        
        return population

    

File: src/population_factory.cpp

#include "population_factory.hpp"

#include <assert.h>

namespace gga
{

    PopulationFactory::PopulationFactory(
        std::vector<std::pair<double, double>> locations,
        std::vector<int>,
        std::vector<int> locationIndices) :
            m_numNodes(locations.size()),
            m_numCells(locationIndices.size()),
            m_population(std::make_shared<Population>(m_numCells))
    {
        for (size_t i = 0; i < m_numCells; i++)
        {
            size_t locationIndex = static_cast<size_t>(locationIndices[i]);
            m_population->cells()[i].location = locations[locationIndex];
        }
    }

    void PopulationFactory::addAges(std::vector<double> ages)
    {
        assert(ages.size() == m_population->cells().size());
        for (size_t i = 0; i < ages.size(); i++)
        {
            m_population->cells()[i].age = ages[i];
        }
    }

    void PopulationFactory::addAncestors(std::vector<int> ancestors)
    {
        assert(ancestors.size() == m_population->cells().size());
        for (size_t i = 0; i < ancestors.size(); i++)
        {
            m_population->cells()[i].ancestor = ancestors[i];
        }
    }

    void PopulationFactory::addAreas(std::vector<double> areas)
    {
        assert(areas.size() == m_population->cells().size());
        for (size_t i = 0; i < areas.size(); i++)
        {
            m_population->cells()[i].area = areas[i];
        }
    }

    void PopulationFactory::addProliferativeTypes(std::vector<ProliferativeCellType> proliferativeTypes)
    {
        assert(proliferativeTypes.size() == m_population->cells().size());
        for (size_t i = 0; i < proliferativeTypes.size(); i++)
        {
            m_population->cells()[i].proliferativeCelltype = proliferativeTypes[i];
        }
    }

    void PopulationFactory::addCellTypes(std::vector<int> cellTypes)
    {
        assert(cellTypes.size() == m_population->cells().size());
        for (size_t i = 0; i < cellTypes.size(); i++)
        {
            m_population->cells()[i].cellType = cellTypes[i];
        }
    }

    PopulationPtr PopulationFactory::population()
    {
        return m_population;
    }


} // namespace gga
